projects/storefrontlib/src/lib/cms-lib/product-carousel/product-carousel.component.ts

```diff
@@ -37,77 +18,49 @@ const SPEED = 250;
 export class ProductCarouselComponent implements OnInit {
   items$: Observable<Observable<Product>[]>;
   itemSize$: Observable<number>;
-  activeItem = 0;
+  activeItem$: Observable<number>;
+  title$: Observable<string>;
 
   private window: Window;
 
   constructor(
-    public component: CmsComponentData<CmsProductCarouselComponent>,
-    private productService: ProductService,
     winRef: WindowRef,
     private el: ElementRef,
-    private cd: ChangeDetectorRef
+    private service: ProductCarouselService
   ) {
     this.window = winRef.nativeWindow;
   }
 
   ngOnInit() {
+    this.setActiveItem(0, 1);
+    this.setTitle();
     this.setItemSize();
     this.setItems();
   }
 
-  /**
-   * Maps the item codes from CMS component to an array of `Product` observables.
-   */
+  protected setTitle(): void {
+    this.title$ = this.service.getTitle();
+  }
   protected setItems(): void {
-    this.items$ = this.component.data$.pipe(
-      map(data => {
-        const productCodes = data.productCodes.split(' ');
-        return productCodes.map(code => this.productService.get(code));
-      })
-    );
+    this.items$ = this.service.getItems();
   }
 
-  /**
-   * The number of items shown in the carousel can be calculated
-   * the standard implemenattions uses the element size to calculate
-   * the items that fit in the carousel.
-   * This method is called in `ngOnInit`.
-   */
   protected setItemSize(): void {
-    if (!this.window) {
-      this.itemSize$ = of(MAX_ITEM_SIZE);
-      return;
-    }
-    this.itemSize$ = fromEvent(this.window, 'resize').pipe(
-      map(() => (this.el.nativeElement as HTMLElement).clientWidth),
-      startWith((this.el.nativeElement as HTMLElement).clientWidth),
-      // avoid to much calls
-      debounceTime(100),
-      map((innerWidth: any) => {
-        const itemsPerPage = Math.round(innerWidth / MAX_WIDTH);
-        return itemsPerPage > 2 ? MAX_ITEM_SIZE : itemsPerPage;
-      }),
-      // only emit new size when the size changed
-      distinctUntilChanged()
+    this.itemSize$ = this.service.getItemSize(
+      this.window,
+      this.el.nativeElement
     );
   }
 
-  prev(max): void {
-    this.setActiveItem(this.activeItem - max, max);
+  prev(activeItem: number, max: number) {
+    this.activeItem$ = this.service.setPreviousItemAsActive(activeItem, max);
```

Comment:
this.activeItem$ = this.service.getPreviousItem(activeItem, max); etc.

Reply:
changed